### PR TITLE
Tidy up hermite_renormalized and add missing gradients

### PR DESCRIPTION
**Pull request by ziofil**
### **User description**
**Context:**
A few lattice methods were missing the gradient, and also there was a proliferation of `hermite_renormalized_something` methods.

**Description of the Change:**
This PR introduces the missing gradients and sets up a single point of entry as `hermite_renormalized` which then calls the appropriate full_batch/b_batch/stable functions.

**Benefits:**
Easier use of hermite_renormalized and most importantly now it's possible to use all of them in training.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Consolidated and refactored `hermite_renormalized` functions for clarity and flexibility.
  - Introduced `hermite_renormalized_unbatched`, `hermite_renormalized_b_batch`, and `hermite_renormalized_full_batch`.
  - Added support for stable and non-stable computation modes.

- Reorganized and modularized the lattice strategies.
  - Moved vanilla strategies into a dedicated module with submodules for core, batch, and gradient functions.
  - Introduced `autoshape_numba` for dynamic Fock representation shape computation.

- Updated backends (NumPy, JAX, TensorFlow) to use the new modular strategies.
  - Added gradient support for batched and unbatched computations.

- Enhanced tests to validate new functionalities and ensure stability.
  - Added tests for stable and batched versions of `hermite_renormalized`.
  - Updated existing tests to use the new function signatures.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>13 files</summary><table>
<tr>
  <td><strong>dm.py</strong><dd><code>Updated import path for `autoshape_numba`</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-17d0cf9d48cb14fcbaadec367030d4f41d5c782c2a069d5d9342569588e2ba38">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ket.py</strong><dd><code>Updated import path for `autoshape_numba`</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-065ae49c98ca8757bb66ef7763a81070e7b0eab2e0899ffebc2b4f2491d6df4e">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>backend_jax.py</strong><dd><code>Added modular `hermite_renormalized` functions with stable mode</code></dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-277d6b9d9077473e0fca45c03a82a1503f283442e0b7503da968e5330ec3c242">+101/-25</a></td>

</tr>

<tr>
  <td><strong>backend_manager.py</strong><dd><code>Centralized `hermite_renormalized` entry point with batching support</code></dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-272a316cdf15707ddb7384f13258135f593fc560e21dc19cf17c16cfb5d0aa1c">+70/-32</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>backend_numpy.py</strong><dd><code>Integrated modular `hermite_renormalized` strategies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-c949c22a107a6cbdc038de07d7b8f9c9c5b3445a876d530ab2ee4780ca00e7a2">+21/-40</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>backend_tensorflow.py</strong><dd><code>Added TensorFlow support for modular <code>hermite_renormalized</code> with <br>gradients</code></dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-9b9d2f0045f871a5f865532687c4b9733e960270cfdb405a5688c2e62568feb0">+51/-37</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>autoshape.py</strong><dd><code>Implemented `autoshape_numba` for dynamic Fock shape computation</code></dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-72927214efc8055ae13ea3a8a4584767c40bb2c30de611f0c8cc2f03690c62fc">+153/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>fast_diagonal.py</strong><dd><code>Updated to use stable and modular lattice strategies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-f4b6d4084408f94a10e5c672c443722b815251240d9e8fed3048d7f6cf40b4a6">+11/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>vanilla.py</strong><dd><code>Removed old vanilla strategies (moved to modular structure)</code></dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-e5930fe6347cb3355e54ff2d78cace247ec1b71bb018410019ac6c56befc98a6">+0/-484</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>__init__.py</strong><dd><code>Added modular vanilla strategies for Fock representation</code>&nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-eaf0264d928f07e560bba34c214781404033b40524c17396b6eb03cc9e4d0aff">+46/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>batch.py</strong><dd><code>Implemented batched vanilla strategies for Fock representation</code></dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-5f357db285994a7c36f9d65ff226664e3e1d8313fbbea13a5e21eb93af0cefcf">+116/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>core.py</strong><dd><code>Added core vanilla strategies for Fock representation</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-6cba18c766796283f626180b60dd270dee13e6707b28a9a5afe4f75033bbdf7e">+195/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>gradients.py</strong><dd><code>Added gradient and VJP strategies for vanilla Fock representation</code></dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-e7efa1f5baaf995600e69824b3048e5dbc3bb688b82b58ebc1696fd51404e91c">+322/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>settings.py</strong><dd><code>Updated stable Fock conversion setting description</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-eaa5467c3f570ae6efabf33215bd7a36eb842dcec72aa1799c3bfb43b6e9f6b7">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>test_fast_diagonal.py</strong><dd><code>Updated test for `fast_diagonal` with stable mode</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-ff440e3fb251b177469f7088bdf0888846c61cb7cf7db77708b34e59a09bc8f4">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test_lattice.py</strong><dd><code>Enhanced tests for modular `hermite_renormalized` strategies</code></dd></td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/574/files#diff-6442ba845586044b7f9f296297deec084fecbdaf8e57f53c44bc7233922f6999">+20/-19</a>&nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: mrmustard/math/lattice/strategies/fast_diagonal.py [hunk 1]
```diff
@@ -28,7 +28,12 @@
 
 
 def fast_diagonal(
-    A: ComplexMatrix, b: ComplexVector, c: complex, output_cutoff: int, pnr_cutoffs: tuple[int, ...]
+    A: ComplexMatrix,
+    b: ComplexVector,
+    c: complex,
+    output_cutoff: int,
+    pnr_cutoffs: tuple[int, ...],
+    stable: bool = False,
```
Comment (apchytr) [inline mrmustard/math/lattice/strategies/fast_diagonal.py:36]: Can you document this arg

File: mrmustard/math/backend_manager.py [hunk 2]
```diff
@@ -676,49 +673,66 @@ def gather(self, array: Tensor, indices: Batch[int], axis: int | None = None) ->
...
-        Returns:
-            The renormalized Hermite polynomial of given shape.
-        """
-        return self._apply("hermite_renormalized", (A, B, C, shape))  # pragma: no cover
+            stable: Whether to use the numerically stable version of the algorithm (also slower).
+
+        Returns:
+            The renormalized Hermite polynomial of given shape preserving the batch dimensions.
+        """
+        stable = stable or settings.STABLE_FOCK_CONVERSION
```
Comment (apchytr) [inline mrmustard/math/backend_manager.py:699]: I think we can avoid this if we just do `stable: bool = settings.STABLE_FOCK_CONVERSION` in the signature

File: mrmustard/math/lattice/strategies/vanilla/batch.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    """
+    batch_size = b.shape[0]
+    G = np.zeros((batch_size,) + shape, dtype=np.complex128)
+    for k in prange(batch_size):
+        G[k] = vanilla_numba(shape, A[k], b[k], c[k])
+    return G
+
+
+@njit(parallel=True)
+def stable_full_batch_numba(shape: tuple[int, ...], A, b, c) -> ComplexTensor:  # pragma: no cover
```
Comment (apchytr) [inline mrmustard/math/lattice/strategies/vanilla/batch.py:49]: Can you remind me why we need to split these out between vanilla and stable? Does njit not allow conditional statements / is it less performant? Just looking at these naively it looks like you could make it a single method with a `stable: bool` arg changing  `vanilla_numba(shape, A[k], b[k], c[k])` to 
```
if stable:
     stable_numba(shape, A[k], b[k], c[k])
else:
     vanilla_numba(shape, A[k], b[k], c[k])
```

File: mrmustard/math/lattice/strategies/vanilla/core.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+# limitations under the License.
+
+"""Core vanilla strategies for Fock representation calculation."""
+
+import numpy as np
+from numba import njit
+
+from mrmustard.utils.typing import ComplexTensor
+
+SQRT = np.sqrt(np.arange(100000))
```
Comment (apchytr) [inline mrmustard/math/lattice/strategies/vanilla/core.py:22]: Can you import from autoshape.py?

**Review by apchytr — APPROVED** [review]
Looks awesome thanks!

File: mrmustard/math/backend_manager.py (hunk 2 above)
Comment (ziofil) [reply mrmustard/math/backend_manager.py:699]: no, that won't work because default values are read only once when the file is scanned

File: mrmustard/math/lattice/strategies/vanilla/batch.py (hunk 3 above)
Comment (ziofil) [reply mrmustard/math/lattice/strategies/vanilla/batch.py:49]: Exactly, it would be less performant because it would need to do the conditional check each time it's called.

Comment (ziofil) [reply mrmustard/math/lattice/strategies/vanilla/batch.py:49]: hmm, but we are doing the check anyways just one caller up...

Comment (ziofil) [reply mrmustard/math/lattice/strategies/vanilla/batch.py:49]: ok, I merged them into one!

File: mrmustard/math/lattice/strategies/vanilla/core.py (hunk 4 above)
Comment (ziofil) [reply mrmustard/math/lattice/strategies/vanilla/core.py:22]: it takes 90 us to run that line, I think it takes longer to resolve the import ^^'

**Review by apchytr — APPROVED** [review]
